File: js/main4.js
// Problem: Dijkstra’s Algorithm with Incorrect Priority Queue Handling
// Points: 10/10
// Expected Output: { A: 0, B: 1, C: 3, D: 6, E: 8 }
class Graph {
    constructor() {
        this.adjList = {};
    }

    addEdge(u, v, w) {
        if (!this.adjList[u]) this.adjList[u] = [];
        if (!this.adjList[v]) this.adjList[v] = [];
        this.adjList[u].push([v, w]);
        this.adjList[v].push([u, w]);  
    }

    dijkstra(start) {
        let dist = {};
        let visited = new Set();
        let queue = [[start, 0]];

        for (let node in this.adjList) {
            dist[node] = Infinity;
        }
        dist[start] = 0;

        while (queue.length > 0) {
            let [node, d] = queue.shift();  

            if (visited.has(node)) continue;
            visited.add(node);

            for (let [neighbor, weight] of this.adjList[node]) {
                let newDist = d + weight;
                if (newDist < dist[neighbor]) {
                    dist[neighbor] = newDist;
                    queue.push([neighbor, newDist]);  
                }
            }
        }
        return dist;
    }
}

let g = new Graph();
g.addEdge("A", "B", 1);
g.addEdge("B", "C", 2);
g.addEdge("C", "D", 3);
g.addEdge("D", "E", 2);
g.addEdge("A", "C", 4);

console.log(g.dijkstra("A"));

File: py/main4.py
# Problem: AVL Tree with Incorrect Rotations
# Points: 10/10
# Expected Output: Balanced AVL Tree
class Node:
    def __init__(self, key):
        self.key = key
        self.left = self.right = None
        self.height = 1

def get_height(node):
    return node.height if node else 0

def balance_factor(node):
    return get_height(node.left) - get_height(node.right)  

def rotate_right(y):
    x = y.left
    y.left = x.right
    x.right = y
    y.height = max(get_height(y.left), get_height(y.right)) + 1
    x.height = max(get_height(x.left), get_height(x.right)) + 1
    return x

def insert(root, key):
    if not root:
        return Node(key)

    if key < root.key:
        root.left = insert(root.left, key)
    else:
        root.right = insert(root.right, key)

    root.height = 1 + max(get_height(root.left), get_height(root.right))

    if balance_factor(root) > 1:
        return rotate_right(root)  

    return root

print(insert(None, 10))
